File: backend/app/services/provider_presets.py
"""常见供应商预设（"添加供应商"时一键填好地址与常见模型）。

**解决的是什么**：登记一个供应商，最会填错的是**接口地址**——各家都不一样，
而且同一家的"原生地址"和"OpenAI 兼容地址"常常不是一个（百炼要加
``/compatible-mode/v1``、Gemini 要加 ``/v1beta/openai/``）。让用户去翻文档抄地址，
是纯粹的摩擦。

**只放事实，不放代码**：这里是一条条"名称 + 地址 + 常见模型 ID"的数据。
模型清单是**便利快照**（取自 models.dev 的公开数据，2026-09），会过时——
所以它只是候选建议，**真正的权威来源仍然是上游探测**（``/providers/{id}/available-models``）：
用户点开「添加模型」时既能选这里的建议，也能选探测回来的，还能直接手写。

**只列 OpenAI 兼容端点**：我们的对话客户端只实现 ``/chat/completions``
（见 ``services/llm.py``）。原生协议（Anthropic Messages、Gemini generateContent）
不在这里——列出来用户配了也用不了，比不列更糟。走兼容层的会在 ``hint`` 里说明。
"""

from __future__ import annotations

from dataclasses import dataclass, field

__all__ = ["PROVIDER_PRESETS", "PresetModel", "ProviderPreset"]


@dataclass(frozen=True, slots=True)
class PresetModel:
    """预设里的一条模型建议。``dim`` 只有向量化模型才有意义。"""

    model_id: str
    label: str = ""
    capabilities: tuple[str, ...] = ("chat",)
    dim: int | None = None


@dataclass(frozen=True, slots=True)
class ProviderPreset:
    """一家常见供应商。``base_url`` 一律是 **OpenAI 兼容**地址。"""

    id: str
    label: str
    kind: str
    base_url: str
    hint: str = ""
    models: tuple[PresetModel, ...] = field(default_factory=tuple)


PROVIDER_PRESETS: tuple[ProviderPreset, ...] = (
    ProviderPreset(
        id="deepseek",
        label="深度求索",
        kind="llm",
        base_url="https://api.deepseek.com",
        hint="密钥：platform.deepseek.com",
        models=(
            PresetModel("deepseek-flash", "DeepSeek Flash"),
            PresetModel("deepseek-v4-pro", "DeepSeek V4 Pro"),
            PresetModel("deepseek-v4-flash", "DeepSeek V4 Flash"),
        ),
    ),
    ProviderPreset(
        id="dashscope",
        label="阿里云百炼",
        kind="llm",
        base_url="https://dashscope.aliyuncs.com/compatible-mode/v1",
        hint="密钥：bailian.console.aliyun.com",
        models=(
            PresetModel("qwen3.8-max", "Qwen3.8 Max"),
            PresetModel("qwen3.8-flash", "Qwen3.8 Flash"),
            PresetModel("qwen3.7-plus", "Qwen3.7 Plus"),
        ),
    ),
    ProviderPreset(
        id="moonshot",
        label="月之暗面",
        kind="llm",
        base_url="https://api.moonshot.cn/v1",
        hint="密钥：platform.moonshot.cn",
        models=(
            PresetModel("kimi-k3", "Kimi K3"),
            PresetModel("kimi-k2.6", "Kimi K2.6"),
        ),
    ),
    ProviderPreset(
        id="zhipu",
        label="智谱 AI",
        kind="llm",
        base_url="https://open.bigmodel.cn/api/paas/v4",
        hint="密钥：open.bigmodel.cn",
        models=(
            PresetModel("glm-5.3", "GLM-5.3"),
            PresetModel("glm-5.3-flash", "GLM-5.3 Flash"),
            PresetModel("glm-5.2", "GLM-5.2"),
        ),
    ),
    ProviderPreset(
        id="siliconflow",
        label="硅基流动",
        kind="llm",
        base_url="https://api.siliconflow.cn/v1",
        hint="密钥：cloud.siliconflow.cn",
        models=(
            PresetModel("deepseek-ai/DeepSeek-V4-Flash", "DeepSeek V4 Flash"),
            PresetModel("zai-org/GLM-5.2", "GLM-5.2"),
            PresetModel("Qwen/Qwen3.6-27B", "Qwen3.6 27B"),
            PresetModel("BAAI/bge-m3", "BGE-M3", ("embedding",), 1024),
            PresetModel("BAAI/bge-reranker-v2-m3", "BGE Reranker v2 M3", ("rerank",), None),
        ),
    ),
    ProviderPreset(
        id="openai",
        label="OpenAI",
        kind="llm",
        base_url="https://api.openai.com/v1",
        hint="密钥：platform.openai.com",
        models=(
            PresetModel("gpt-5.6", "GPT-5.6"),
            PresetModel("gpt-5.5", "GPT-5.5"),
            PresetModel("text-embedding-3-small", "Embedding 3 Small", ("embedding",), 1536),
            PresetModel("text-embedding-3-large", "Embedding 3 Large", ("embedding",), 3072),
        ),
    ),
    ProviderPreset(
        id="anthropic",
        label="Anthropic",
        kind="llm",
        base_url="https://api.anthropic.com/v1",
        hint="密钥：console.anthropic.com",
        models=(
            PresetModel("claude-opus-4-8", "Claude Opus 4.8"),
            PresetModel("claude-sonnet-4-6", "Claude Sonnet 4.6"),
        ),
    ),
    ProviderPreset(
        id="gemini",
        label="Google Gemini",
        kind="llm",
        base_url="https://generativelanguage.googleapis.com/v1beta/openai/",
        hint="密钥：aistudio.google.com",
        models=(
            PresetModel("gemini-3.8-flash", "Gemini 3.8 Flash"),
            PresetModel("gemini-3.7-flash", "Gemini 3.7 Flash"),
        ),
    ),
    ProviderPreset(
        id="xai",
        label="xAI",
        kind="llm",
        base_url="https://api.x.ai/v1",
        hint="密钥：console.x.ai",
        models=(
            PresetModel("grok-4.6", "Grok 4.6"),
            PresetModel("grok-4.5", "Grok 4.5"),
        ),
    ),
    ProviderPreset(
        id="ollama",
        label="Ollama",
        kind="llm",
        base_url="http://localhost:11434/v1",
    ),
    # 「自定义」不在这里：选择器本身就有一项「自定义」，再加一个同义条目只是噪声。
)

File: frontend/src/components/settings/providerPresets.ts
/**
 * 供应商预设的纯逻辑（与 Vue 无关，便于单测）。
 *
 * 预设解决的是"接口地址各家不一样"：百炼要 `/compatible-mode/v1`、
 * Gemini 要 `/v1beta/openai/`、自建的千奇百怪。抄地址是纯摩擦，而且抄错了
 * 报错在很后面（探活失败、甚至对话时才 404）。这里只做三件事：
 *
 * 1. 把地址规范化后比较（尾部斜杠不算不同）；
 * 2. 按地址把供应商和预设对上；
 * 3. 把"上游探测到的模型"与"预设建议的模型"合并去重。
 *
 * 预设本身来自后端（`registry.provider_presets`），这里不内置任何地址清单。
 */

import type { PresetModel, Provider, ProviderPreset } from '@/api/modelRegistry'

/** 尾部斜杠归一化：`https://x/v1/` 与 `https://x/v1` 是同一个地址。 */
export function normalizeUrl(url: string): string {
  return url.trim().replace(/\/+$/, '')
}

/**
 * 这个供应商对应哪个预设。
 *
 * 按**地址**匹配而不是按名称：用户可能把"深度求索"改成"我的 DeepSeek"，
 * 地址才是那件真正决定"发到哪去"的事。改过地址就匹配不上——也就不给建议，
 * 好过拿一个不对的预设列表去误导。
 */
export function presetForProvider(
  presets: ProviderPreset[],
  provider: Pick<Provider, 'base_url'>,
): ProviderPreset | null {
  if (!provider.base_url) return null
  const target = normalizeUrl(provider.base_url)
  return (
    presets.find((preset) => preset.base_url && normalizeUrl(preset.base_url) === target) ?? null
  )
}

export function suggestedModels(
  presets: ProviderPreset[],
  provider: Pick<Provider, 'base_url'>,
): PresetModel[] {
  return presetForProvider(presets, provider)?.models ?? []
}

/**
 * 「添加模型」下拉的候选：上游探测到的 + 预设建议（去重）。
 *
 * 探测是**权威**（反映上游此刻真有什么），预设是**兜底**（有的端点不返回模型
 * 列表；新模型上游还没上架时也有个写法可参考）。预设项打 `· 预设` 标记，
 * 让用户知道这条不是探测回来的。
 */
export function mergeModelOptions(
  probeOptions: { value: string; label: string }[],
  presets: ProviderPreset[],
  provider: Pick<Provider, 'base_url'>,
): { value: string; label: string }[] {
  const known = new Set(probeOptions.map((item) => item.value))
  const suggested = suggestedModels(presets, provider)
    .filter((item) => !known.has(item.model_id))
    .map((item) => ({ value: item.model_id, label: `${item.label || item.model_id} · 预设` }))
  return [...probeOptions, ...suggested]
}

/**
 * 选中预设建议里的某条模型时，它自带的元数据（能力、维度）。
 *
 * **只对预设项返回**：手写或探测到的 ID 没有元数据，猜能力等于替用户做决定
 * （把 embedding 标成 chat，要到建库时才发现）。
 */
export function presetModelMeta(
  presets: ProviderPreset[],
  provider: Pick<Provider, 'base_url'>,
  modelId: string,
): PresetModel | null {
  const value = modelId.trim()
  if (!value) return null
  return suggestedModels(presets, provider).find((item) => item.model_id === value) ?? null
}
